### ISSUE TITLE
zram-based swap integration

### ISSUE THREAD

**Issue by cmurf** [cockpit-project/cockpit#14436]
Page:  Overview > Usage. Maybe Storage, but swap is less about storage per se, than it is about memory management, especially when zram-based.

Fedora 33, all editions and spins, will come with zram-based swap by default instead of disk-based swap. It might be useful to indicate this somewhere, possibly with minimal UI to just change the size.
https://fedoraproject.org/wiki/Changes/SwapOnZRAM

There might also need to be some deconfliction with udisks2-zram module (?) as discovered later in this bug:
https://bugzilla.redhat.com/show_bug.cgi?id=1861463#c9

Should the interface make it easy to completely disable zram-based swap? I'll argue no. 

1. There's an advantage of always having some minimal amount because it will reduce reclaim. Reclaim can be about as expensive as swap thrashing in terms of IO and latency. Since the zram device is not a preallocation of RAM, it does no harm to have it just sit there unused, but ready to permit eviction of inactive pages via compression rather than being able to depend only on reclaim. 

2. A minimum swap ties in better with earlyoom today, and in the near future with systemd-oomd, both of which evaluate swap usage to make assessment.

3. tmpfs depends on swap for overflow when physical memory pressure demands it.

What's a reasonable minimum? 512M?

Should the UI be fraction or absolute size based?  The zram-generator.conf offers `zram-fraction=` and `max-zram-size=` which could be exposed directly to to the user. Or just have them specify an absolute size, and translate that into appropriate values behind the scene.

**Comment by garrett** [comment]
**Summary** of the below: zram is great. I'm personally a bit hesitant to expose zram-specific tuning in Cockpit, however. (I could be convinced, though.) But, regardless, we should consider what we show & do with zram in Cockpit.

---

Just chiming in to state that I recently switched my machines over to using zram-based swap and it's wonderful. It's made my Fedora-based desktops more responsive, even with their huge amounts of RAM. (I often would hit a point most days where the machine would get stuck with IO for several seconds to minutes. This hasn't happened since the switch.) It's helped my Raspberry Pi servers quite a bit too (Pi 4 & Pi 1 especially).

I'm not sure if people will generally use zram (or already doing so) on servers. However, with Fedora's upcoming adoption of it, I imagine at least Fedora-based servers will use it. And I wouldn't be surprised to see other distros supporting it too.

We have talked a little bit about zram-based swap and how that should be shown in Cockpit. But it was limited to performance metrics. Our conclusion was that if a machine is swapping significantly, even using zram swap instead of disk swap, that means there's memory pressure and we should show that.

There's a little bit of CPU overhead for zram, but it's much less than IO-related disk/SSD thrashing. (And this, of course, should make it appealing for servers too.)

I'm not sure if Cockpit should expose a modifiable value for zram-based swap, however, due to the nature if it not taking up much space when it's not in use. Hopefully Fedora will have reasonable defaults.

From what I've set up, I have up to half the amount of RAM in a system as a maximum for the swap space. It's worked well (and I haven't really even approached it). FWIW: Fedora has adjusted their recommendation for their [default to be 50% of RAM up until 4 GB](https://fedoraproject.org/wiki/Changes/SwapOnZRAM#Default_zram_device_configuration:).

**Comment by martinpitt** [comment]
I agree that we shouldn't expose any zram tuning knobs -- few people should have a need to change that, this is an OS builder's job. @cmurf, so what exactly is being asked here? I.e. what kind of functionality do you think cockpit should have wrt. zram?
